rllib/core/optim/rl_optimizer.py

```diff
@@ -0,0 +1,94 @@
+import abc
+from typing import Any, List, Mapping
+
+from ray.rllib.core.rl_module import RLModule
+from ray.rllib.utils.annotations import (
+    OverrideToImplementCustomLogic,
+    OverrideToImplementCustomLogic_CallToSuperRecommended,
+)
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI(stability="beta")
+class RLOptimizer(abc.ABC):
+    """Base class for defining a loss function and optimizer for a RLModule.
+
+    Args:
+        rl_module: The RLModule that will be optimized.
+        config: The configuration for the optimizer.
+
+    Abstract Methods:
+        compute_loss: computing a loss to optimize rl_module over.
+        _configure_optimizers: constructing the optimizers for rl_module.
+        get_state: getting the state of the optimizer.
+        set_state: setting the state of the optimizer.
+
+    Example:
+    .. code-block:: python
+
+        module = RLModule(...)
+        rl_optim = RLOptimizer(module, config)
+        sample_batch = ...
+        fwd_out = module.forward_train(sample_batch)
+        loss_dict = rl_optim.compute_loss(fwd_out)
+
+        # compute gradients of loss w.r.t. trainable variables
+        ...
+
+        for optim in rl_optim.get_optimizers():
+            optim.step()
+            optim.zero_grad()
+
+    """
+
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    def __init__(self, module: RLModule, config: Mapping[str, Any]):
+        self.module = module
+        self._config = config
+        self._optimizers = self._configure_optimizers()
+
+    @abc.abstractmethod
+    def compute_loss(self, fwd_out: Mapping[str, Any]) -> Mapping[str, Any]:
+        """Computes variables for optimizing self._module based on fwd_out.
+
+        Args:
+            fwd_out: Output from a forward pass on self._module during
+                training.
+
+        Returns:
+            A dictionary of tensors used for optimizing self._module.
+        """
+
+    @OverrideToImplementCustomLogic
+    @staticmethod
+    def on_after_compute_loss(loss_dict: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
question: What are some concrete examples of this function in action?

Reply:
doing something like loss clipping, or adding additional losses without modifying the existing loss function. for example adding a penalty for the norm of the network to keep weights small, or entropy penalty. you could do these in the regular compute loss fn, but if you were doing fast development the hook might be more convenient to you

Reply:
I see. So this function has no special functionality privilege over compute_loss? I am particularly looking for use-cases where having this abstraction of after_loss and the loss itself provides significant value that otherwise would have been very difficult. I think if all the cases that you can think of can be handled by overriding or extending compute_loss we should just remove the after_loss hook.